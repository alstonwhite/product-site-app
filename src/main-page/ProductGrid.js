import React from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link,
  useHistory,
  useLocation,
  useParams
} from "react-router-dom";

import ProductTile from './ProductTile'

const ProductGrid = ({products}) => {

  const lookup = [
    {
      display: "Product Group 1",
      kebab: "product-group-1"
    },
    {
      display: "Product Group 2",
      kebab: "product-group-2"
    },
    {
      display: "Product Group 3",
      kebab: "product-group-3"
    },
    {
      display: "Product Group 4",
      kebab: "product-group-4"
    }
  ]
  // add slug (kebab) to testData, lookup against that array

  let filter = useParams().group;
  let filterDisplay = filter ?
    lookup.find(x => x.kebab === filter).display :
    "All Products";


  return (
    <div className="product-grid" id="product-grid">
      <h1>{filterDisplay}</h1>
      <div className="product-grid-items">
          {filter ?
            products.map(product => (
              product.section === filterDisplay &&
              <ProductTile 
                  product = {product}
              />))
            : products.map(product => (
              <ProductTile 
                  product = {product}
              />
          ))}
      </div>
    </div>
  );
};
  
  export default ProductGrid;