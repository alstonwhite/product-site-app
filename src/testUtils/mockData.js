export const testSections = 
    [
        {
            id: "0",
            slug: "section1",
            sectionTitle: "Section 1",
            active: true
        },
        {
            id: "1",
            slug: "section2",
            sectionTitle: "Section 2",
            active: true
        }
    ]

export const testProducts = 
    [
        {
            id: "0",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 1",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "1",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 2",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "2",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 3",
            description: "Description",
            productSection: {fields: {slug: "section2"}},
            active: true
        }
    ]

export const testProduct = 
    {
        id: 0,
        image1: {fields: {file: {url: "1"}}},
        image2: {fields: {file: {url: "2"}}},
        image3: {fields: {file: {url: "3"}}},
        price: 5,
        title: "Product 1",
        description: "Description",
        productSection: {fields: {slug: "section1"}},
        active: true
    }

export const testCartProduct = 
[{
    id: 0,
    image1: {fields: {file: {url: "1"}}},
    image2: {fields: {file: {url: "2"}}},
    image3: {fields: {file: {url: "3"}}},
    price: 5,
    title: "Product 1",
    description: "Description",
    productSection: {fields: {slug: "section1"}},
    active: true,
    quantity: 1
}]
