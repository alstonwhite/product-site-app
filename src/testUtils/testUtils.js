import React from "react";
import { Router, Route } from "react-router-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";

export const renderWithRouter = (ui, 
    {
        path="/",
        route="/",
        history = createMemoryHistory({ initialEntries: [route] }) 
    }={}) => ( 
    render( 
        <Router history={history}> 
            <Route path={path} children={ui} /> 
        </Router> 
    ) 
)