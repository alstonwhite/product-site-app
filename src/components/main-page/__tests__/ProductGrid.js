import React from 'react'
import { cleanup, screen } from '@testing-library/react'
import ProductGrid from '../ProductGrid'
import { testProducts } from '../../../testUtils/mockData'
import { renderWithRouter } from '../../../testUtils/testUtils'


afterEach(cleanup)

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
    useDispatch: () => mockDispatch
}));

test('renders proper product tiles when unfiltered', () => {
    renderWithRouter(<ProductGrid products={testProducts}/>, {path: "/"})
    
    expect(screen.getAllByTestId('product-tile')).toHaveLength(testProducts.length)
})

test('renders proper product tiles when filtered', () => {
    const filter = "section1"
    renderWithRouter(<ProductGrid products={testProducts}/>, 
        {path: "/category/:group", route: `/category/${filter}`})
    
        expect(screen.getAllByTestId('product-tile')).toHaveLength(
        testProducts.filter(p => p.productSection.fields.slug === "section1").length
        )
})
  
test('snapshot', () => {
    const {container} = renderWithRouter(<ProductGrid products={testProducts}/>, {path: "/"})

    expect(container.firstChild).toMatchSnapshot()
})

