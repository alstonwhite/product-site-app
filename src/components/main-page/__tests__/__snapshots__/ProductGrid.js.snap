// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="product-grid"
  id="product-grid"
>
  <h1
    class="product-grid__header"
  >
    All Products
  </h1>
  <div
    class="product-grid__items"
  >
    <div
      class="product-tile"
      data-testid="product-tile"
      id="product-tile"
    >
      <a
        class="product-tile__content"
        href="/product/0"
      >
        <img
          alt="logo"
          class="product-tile__img"
          src="1"
        />
        <div
          class="product-tile__name"
        >
          Product 1
        </div>
        <div
          class="product-tile__price"
        >
          <span>
            $
          </span>
          5
        </div>
      </a>
    </div>
    <div
      class="product-tile"
      data-testid="product-tile"
      id="product-tile"
    >
      <a
        class="product-tile__content"
        href="/product/1"
      >
        <img
          alt="logo"
          class="product-tile__img"
          src="1"
        />
        <div
          class="product-tile__name"
        >
          Product 2
        </div>
        <div
          class="product-tile__price"
        >
          <span>
            $
          </span>
          5
        </div>
      </a>
    </div>
    <div
      class="product-tile"
      data-testid="product-tile"
      id="product-tile"
    >
      <a
        class="product-tile__content"
        href="/product/2"
      >
        <img
          alt="logo"
          class="product-tile__img"
          src="1"
        />
        <div
          class="product-tile__name"
        >
          Product 3
        </div>
        <div
          class="product-tile__price"
        >
          <span>
            $
          </span>
          5
        </div>
      </a>
    </div>
  </div>
</div>
`;
